### PR TITLE
[SPARK-5205][Streaming]:Inconsistent behaviour between Streaming job and others, when click kill link in WebUI

### PR DESCRIPTION
**Pull request by uncleGen**
The "kill" link is used to kill a stage in job. It works in any kinds of Spark job but Spark Streaming. To be specific, we can only kill the stage which is used to run "Receiver", but not kill the "Receivers". Well, the stage can be killed and cleaned from the ui, but the receivers are still alive and receiving data. I think it dose not fit with the common sense. IMHO, killing the "receiver" stage means kill the "receivers" and stopping receiving data.

### CODE REVIEW

**Comment by tdas** [comment]
@pwendell This should be a good addition to TaskContext, take a look.

**Comment by uncleGen** [comment]
@JoshRosen Could you please take a look again, thank you.

**Comment by uncleGen** [comment]
@JoshRosen Could you please take a look again, thank you!

**Comment by uncleGen** [comment]
@pwendell Could you please take a look again, thank you!

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
I think we discussed this on the first PR, but it would be good to add an explicit note here summarizing why we're not using `TaskCompletionListener`, since its documentation says that it's called on all task completion events, including success, failure, or cancellation.

**Comment by JoshRosen** [comment]
Actually, this just occurred to me: why not use `addTaskCompletionListener` and check `TaskContext.isInterrupted()` inside of your listener in order to decide whether to perform the receiver shutdown?  If this works, then this patch becomes a two-line change to ReceiverTracker.scala:

```scala
context.addTaskCompletionListener { context =>
  if (context.isInterrupted()) {
    supervisor.stop("Receiver was killed.", None)
  }
}

**Comment by JoshRosen** [comment]
Can you try using the existing `addTaskCompletionListener` with `context.isInterrupted()`, roll back the other changes / listener interfaces, and add a streaming unit test for this receiver cancellation behavior?

**Comment by uncleGen** [comment]
retest this please

**Comment by uncleGen** [comment]
Some timeout errors happened. Could you please take a time to check and review it again?

**Comment by uncleGen** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/TaskContextImpl.scala [hunk 1]
```diff
@@ -66,8 +66,21 @@ private[spark] class TaskContextImpl(
   /** Marks the task as completed and triggers the listeners. */
   private[spark] def markTaskCompleted(): Unit = {
     completed = true
+    processCallBacks()
+  }
+
+  /** Marks the task for interruption, i.e. cancellation. */
+  private[spark] def markInterrupted(): Unit = {
+    interrupted = true
+    processCallBacks()
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/TaskContextImpl.scala]: I noticed that the old code did not process callbacks on interruption.  What is the API contract here for markTaskCompleted and markInterrupted()?  Are they mutually-exclusive, e.g. a task can be marked as completed or interrupted, but never both?  I think that we should add assertions to determine if this is the case, since I'm worried that this change might lead to duplicate calling of callbacks if this assumption isn't true.

Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/TaskContextImpl.scala]: Actually, tale a look at the definition of `Task.run`:

```scala
  final def run(attemptId: Long): T = {
    context = new TaskContextImpl(stageId, partitionId, attemptId, false)
    TaskContextHelper.setTaskContext(context)
    context.taskMetrics.hostname = Utils.localHostName()
    taskThread = Thread.currentThread()
    if (_killed) {
      kill(interruptThread = false)
    }
    try {
      runTask(context)
    } finally {
      context.markTaskCompleted()
      TaskContextHelper.unset()
    }
  }
```

If `_killed` is true, then we'll call `kill()` which marks the task context as interrupted, and then still call `runTask`, so the `finally` block will mark the task as completed after it's already been interrupted.  Therefore, I don't think that triggering call backprocessing inside of `markInterrupted` is the right approach, especially since `Task.kill()` is asynchronous, so the task could still be running at this point.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala [hunk 2]
```diff
@@ -23,7 +23,9 @@ import scala.language.existentials
 
 import akka.actor._
 
-import org.apache.spark.{Logging, SerializableWritable, SparkEnv, SparkException}
+import scala.Some
```
Comment (JoshRosen) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala]: This import should be grouped with the other Scala imports on line 22.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala [hunk 3]
```diff
@@ -295,6 +297,14 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
         val receiver = iterator.next()
         val supervisor = new ReceiverSupervisorImpl(
           receiver, SparkEnv.get, serializableHadoopConf.value, checkpointDirOption)
+
+        // Due to SPARK-5205, `Receiver` stage can not be stopped properly, we need to add a
+        // callback to do some more clean works.
+        context.addTaskCompletionListener { context =>
+          if(context.isInterrupted()) {
```
Comment (JoshRosen) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala]: Code style: space after `if`.

File: streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala [hunk 4]
```diff
@@ -129,6 +130,41 @@ class ReceiverSuite extends TestSuiteBase with Timeouts with Serializable {
...
+      val receiver = new FakeReceiver
+      val executor = new FakeReceiverSupervisor(receiver)
+      val task = new FakeReceiverTask(0, executor)
+
+      new Thread("Receiver Task") {
+        override def run(): Unit = {
+          task.run(0, 0)
+        }
+      }.start()
+      Thread.sleep(2000)
```
Comment (JoshRosen) [inline streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala]: If we can help it, we shouldn't be writing any new streaming tests that rely on Thread.sleep, since this has been a major cause of test flakiness.  See the use of ManualClock in the other streaming tests for examples of how to write tests which aren't dependent on wall-clock time.

File: streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala [hunk 5]
```diff
@@ -129,6 +130,41 @@ class ReceiverSuite extends TestSuiteBase with Timeouts with Serializable {
...
+
+      // Verify that receiver was started
+      assert(receiver.onStartCalled)
+      assert(executor.isReceiverStarted)
+      assert(receiver.isStarted)
+      assert(!receiver.isStopped())
+      assert(receiver.otherThread.isAlive)
+
+      task.kill(false)
+
```
Comment (JoshRosen) [inline streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala]: There's some asynchrony in task killing, so these checks that the receiver was stopped need to be wrapped in an `eventually` block or some use some other mechanism to avoid race conditions.

Comment (JoshRosen) [inline streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala]: In fact, there's an example of using `eventually` in this file; see line 112.

File: streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala [hunk 6]
```diff
@@ -296,6 +332,27 @@ class ReceiverSuite extends TestSuiteBase with Timeouts with Serializable {
...
+   * An implementation of Task running a NetworkReceiver used for testing whether
+   * we can start/stop receiver properly.
+   */
+  class FakeReceiverTask(
+    stageId: Int,
+    executor: FakeReceiverSupervisor,
+    prefLocs: Seq[TaskLocation] = Nil) extends Task[Unit](stageId, 0) {
+    override def runTask(context: TaskContext): Unit = {
+      context.addTaskCompletionListener { context =>
+        if(context.isInterrupted()) {
```
Comment (JoshRosen) [inline streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala]: Style nit: space after `if`.

**Comment by uncleGen** [comment]
@JoshRosen, thanks for your patience. It occurred to me that we may check when to terminate the `receiver` in `ReceiverSupervisor`. Then the condition to stop the `receiver` is satisfied when `TaskContext.isInterrupted()` or `isReceiverStopped`. So I removed the `stopLatch` lock and just `checkTermination()` once per 500ms. Well, a little ugly and any idea?

Now, we may just mark the task for interruption and trigger callback processing inside of markTaskCompleted.

File: streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala [hunk 7]
```diff
@@ -161,22 +163,45 @@ private[streaming] abstract class ReceiverSupervisor(
...
 
-  /** Check if receiver has been marked for stopping */
+  /** Check if receiver has been marked for stopping. */
   def isReceiverStopped() = {
     logDebug("state = " + receiverState)
     receiverState == Stopped
   }
 
+  /** Due to SPARK-5205, we need `TaskContext` to check if the task is interrupted. */
+  def setTaskContext(context: TaskContext) = {
```
Comment (JoshRosen) [inline streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala]: Instead of adding a setter method, why not pass the TaskContext into the ReceiverSupervistor's constructor?

**Comment by JoshRosen** [comment]
I chatted with @tdas about this and I think we've arrived at a better understanding of this problem:

- The `onTaskCompletion` approach that I suggested will not work because that callback will only be invoked after the task has finished running, but the task will not be able to finish until we stop the receiver.
- We could modify the receiver to respond to `TaskContext.isInterrupted()`, but the current implementation here isn't ideal since it involves busy-looping.

It sounds like your original approach of adding a new callback on interruption is the right fix, since that callback can just call `Receiver.stop()` and we can continue to use the countdown latch to avoid polling / busy-waiting.  However, I think that the original name for the callback was misleading / confusing, which led me onto this derail: the name `addTaskKilledListener` is confusing because it sounds like it would be subsumed by `onCompleteCallback`.  Besides this naming confusion, though, I think the original approach is actually fine.

Therefore, my suggestion is to keep the unit tests that you've added (which look good to me) and revert back to the `onKilledCallback` approach, but change the name to be `addTaskInterruptionListener`.  Because the existing methods / naming was confusing, it might also be nice to add a state transition diagram somewhere which explains the order in which the callbacks are invoked (or some much more explicit comments).

Sorry for the derail and thanks for being patient with me!

**Comment by uncleGen** [comment]
@JoshRosen OK, I will roll back to the original approach and do some impromvements :)

**Comment by uncleGen** [comment]
close and resolve the merge conflict.

**Comment by uncleGen** [comment]
@JoshRosen Could you please take a look again, thank you!

File: core/src/main/scala/org/apache/spark/TaskContext.scala [hunk 8]
```diff
@@ -106,6 +105,20 @@ abstract class TaskContext extends Serializable {
   def addOnCompleteCallback(f: () => Unit)
 
   /**
+   * Add a (Java friendly) listener to be executed on task interruption. We add this
+   * listener for some more clean works. An example use is to stop `receiver supervisor`
+   * properly.
+   */
+  def addTaskInterruptedListener(listener: TaskInterruptionListener): TaskContext
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/TaskContext.scala]: For consistency, I think we should call this `addTaskInterruptionListener`, since the actual listener class is named `TaskInterruptionListener` and for consistency with the other listener, which is named "completion" rather than "complete".

File: core/src/main/scala/org/apache/spark/TaskContextImpl.scala [hunk 9]
```diff
@@ -82,9 +97,27 @@ private[spark] class TaskContextImpl(
     }
   }
 
-  /** Marks the task for interruption, i.e. cancellation. */
+  /**
+   * Marks the task as interruption, i.e. cancellation. We add this
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/TaskContextImpl.scala]: I'd reword this comment slightly to say something like "Marks the task for interruption (i.e. cancellation) and calls any task interruption callbacks registered for this task."

We already document the sample use-case on the public addTaskInterruptionListener callback, so I don't think we need to do it here as well.

File: core/src/main/scala/org/apache/spark/TaskContextImpl.scala [hunk 10]
```diff
@@ -82,9 +97,27 @@ private[spark] class TaskContextImpl(
...
+   */
   private[spark] def markInterrupted(): Unit = {
     interrupted = true
+    val errorMsgs = new ArrayBuffer[String](2)
+    // Process interruption callbacks in the reverse order of registration
+    onInterruptedCallbacks.reverse.foreach { listener =>
+      try {
+        listener.onTaskInterrupted(this)
+      } catch {
+        case e: Throwable =>
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/TaskContextImpl.scala]: It might be dangerous for this code to catch all Throwables, since this could also catch irrecoverable exceptions like OutOfMemoryError.  Instead, it might be safer to use `case NonFatal(e) =>`.  If you agree, then let's also update the original code in `markTaskCompleted()` to catch NonFatal as well.

File: streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala [hunk 11]
```diff
@@ -23,9 +23,9 @@ import scala.collection.mutable.ArrayBuffer
 
 import org.apache.spark.{Logging, SparkConf}
 import org.apache.spark.storage.StreamBlockId
-import java.util.concurrent.CountDownLatch
 import scala.concurrent._
 import ExecutionContext.Implicits.global
+import java.util.concurrent.CountDownLatch
```
Comment (JoshRosen) [inline streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala]: The import ordering in this file was a bit of a mess to begin with, but if you're going to change it would you mind bringing it up to date with the ordering described in the style guide (e.g. java imports, then Scala, then third-party packages, then Spark)?  https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide

Comment (JoshRosen) [inline streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala]: We don't actually have to modify this file as part of this patch, though, so it'd be fine to just roll it back to what's in master.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala [hunk 12]
```diff
@@ -295,6 +296,14 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
         val receiver = iterator.next()
         val supervisor = new ReceiverSupervisorImpl(
           receiver, SparkEnv.get, serializableHadoopConf.value, checkpointDirOption)
+
+        // Due to SPARK-5205, `Receiver` stage can not be stopped properly, we need to add a
```
Comment (JoshRosen) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala]: It's fine to be a bit more explicit / verbose about what's going on here, so I'd maybe update this comment to say something like this:

> To ensure that receivers are stopped properly when their tasks / stages are killed, register a callback to stop the receiver when its task is interrupted.  See SPARK-5205 for more details.

File: streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala [hunk 13]
```diff
@@ -296,6 +335,27 @@ class ReceiverSuite extends TestSuiteBase with Timeouts with Serializable {
...
   /**
+   * An implementation of Task running a NetworkReceiver used for testing whether
+   * we can start/stop receiver properly.
+   */
+  class FakeReceiverTask(
+    stageId: Int,
+    executor: FakeReceiverSupervisor,
+    prefLocs: Seq[TaskLocation] = Nil) extends Task[Unit](stageId, 0) {
+    override def runTask(context: TaskContext): Unit = {
+      context.addTaskInterruptedListener { context =>
```
Comment (JoshRosen) [inline streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala]: I notice that this test harness the logic from the real receiver tracker.  Is this test suite actually covering the changes made to ReceiverTracker?  In other words, if I were to go in and delete the callback registration code that you added to the real tracker, would that cause any tests to fail?  If not, then I think we need to do additional work to test this, since the test isn't really testing anything if it can't expose bugs in the real code.

It seems like we'd like to test against a real ReceiverTracker using mock implementations of ReceiverSupervisor and Receiver.  I'm not super-familiar with this corner of the streaming code, though, so maybe @tdas can chime in with suggestions on how to test this?

**Comment by JoshRosen** [comment]
I left a pass of comments, mostly about naming / documentation things.  The overall approach looks good, but I'm not sure if this is the right / best unit test for this logic because the test harness seems to duplicate most of the logic being tested.  I'll let @tdas chime in regarding tests.

**Comment by JoshRosen** [comment]
Any update on this?  This patch looked good to me overall; the main thrust of my feedback was that the tests here didn't seem sufficient to actually exercise the code added in this patch, but I think that shouldn't be too hard to fix.

**Comment by uncleGen** [comment]
@JoshRosen Sorry for my laziness. I will update it as soon as possible

**Comment by JoshRosen** [comment]
@uncleGen No worries; I'm not in a huge rush, but just wanted to check in to see if you planned to update this.  Sorry that the review process has been so slow; thanks for your patience!

**Comment by uncleGen** [comment]
@JoshRosen Your comments are reasonable, and I have improved related code just as what you pointed. For the test suite, I just check if the state of `Receiver` and `ReceiverSupervisor` are correct. In the previous way, killing `task` could not stop `Receiver` and `ReceiverSupervisor` properly. I think it is enough. @tdas, what is your opinion?

**Comment by uncleGen** [comment]
wait for a moment

**Comment by uncleGen** [comment]
@JoshRosen Your comments are reasonable, and I have improved the relevant codes just what you pointed. And more, I add some unit tests about the  behavior about `TaskInterruptionListener`. In the same time, I left some unit test in `ReceiverSuite` to test the behavior about `Receiver` and `ReceiverSupervisor`. Maybe, @tdas has some suggestions.

**Comment by uncleGen** [comment]
Ha~llo~, @JoshRosen

File: core/src/main/scala/org/apache/spark/util/TaskInterruptionListener.scala [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+
+import java.util.EventListener
+
+import org.apache.spark.TaskContext
+import org.apache.spark.annotation.DeveloperApi
+
+/**
+ * :: DeveloperApi ::
+ *
+ * Listener providing a callback function to invoke when interrupt a task.
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/util/TaskInterruptionListener.scala]: "when interrupt" -> "when interrupting"

File: core/src/main/scala/org/apache/spark/TaskContext.scala [hunk 15]
```diff
@@ -106,6 +105,20 @@ abstract class TaskContext extends Serializable {
...
   /**
+   * Add a (Java friendly) listener to be executed on task interruption. We add this
+   * listener for some more clean works. An example use is to stop `receiver supervisor`
+   * properly.
+   */
+  def addTaskInterruptionListener(listener: TaskInterruptionListener): TaskContext
+
+  /**
+   * Add a listener in the form of a Scala closure to be executed on task interruption.
+   * We add this listener for some more clean works. An example use is to stop `receiver
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/TaskContext.scala]: I'd reword this slightly to be a bit more explicit:

> Add a listener in the form of a Scala closure to be executed on task interruption.  An example use is to cleanly shut down long-running receiver tasks in Spark Streaming.

File: core/src/test/scala/org/apache/spark/scheduler/TaskContextSuite.scala [hunk 16]
```diff
@@ -64,6 +64,53 @@ class TaskContextSuite extends FunSuite with BeforeAndAfter with LocalSparkConte
...
+        Thread.sleep(1000)
+        Iterator("test")
+      }
+    }
+    val closureSerializer = SparkEnv.get.closureSerializer.newInstance()
+    val func = (c: TaskContext, i: Iterator[String]) => i.next()
+    val task = new ResultTask[String, String](
+      0, sc.broadcast(closureSerializer.serialize((rdd, func)).array), rdd.partitions(0), Seq(), 0)
+
+    task.kill(false)
```
Comment (JoshRosen) [inline core/src/test/scala/org/apache/spark/scheduler/TaskContextSuite.scala]: The task isn't running at this point, so what's the purpose of calling `kill` here?

File: core/src/test/scala/org/apache/spark/scheduler/TaskContextSuite.scala [hunk 17]
```diff
@@ -64,6 +64,53 @@ class TaskContextSuite extends FunSuite with BeforeAndAfter with LocalSparkConte
...
+    assert(TaskContextSuite.interrupted === false)
+
+    val taskThread = new Thread("Task") {
+      override def run(): Unit = {
+        task.run(0, 0)
+      }
+    }
+    taskThread.start()
+    Thread.sleep(500)
```
Comment (JoshRosen) [inline core/src/test/scala/org/apache/spark/scheduler/TaskContextSuite.scala]: If possible, we should avoid adding any new Thread.sleep() calls in our tests, since they slow things down and have been a major source of flakiness in the past.  How about adding a CountdownLatch to the TaskContexSuite companion object and using that to block the main thread until the body of the task has started executing and the interruption listener has been registered?

File: project/MimaExcludes.scala [hunk 18]
```diff
@@ -64,6 +64,10 @@ object MimaExcludes {
             // SPARK-6492 Fix deadlock in SparkContext.stop()
             ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.SparkContext.org$" +
                 "apache$spark$SparkContext$$SPARK_CONTEXT_CONSTRUCTOR_LOCK")
+		  ) ++ Seq(
```
Comment (JoshRosen) [inline project/MimaExcludes.scala]: The indentation here looks a bit off.  Mind indenting this back a bit so that it lines up with the other closing braces (e.g. line 63)?

File: streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala [hunk 19]
```diff
@@ -129,6 +130,44 @@ class ReceiverSuite extends TestSuiteBase with Timeouts with Serializable {
     }
   }
 
+  test("check the state of `Receiver` and `ReceiverSupervisor` when receivers are killed") {
```
Comment (JoshRosen) [inline streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala]: I'd add `(SPARK-5205)` to the name of this test, if it will fit, or just add a comment on the next line referencing the JIRA number.

File: streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala [hunk 20]
```diff
@@ -295,6 +334,27 @@ class ReceiverSuite extends TestSuiteBase with Timeouts with Serializable {
   }
 
   /**
+   * An implementation of Task running a NetworkReceiver used for testing whether
+   * we can start/stop receiver properly.
+   */
+  class FakeReceiverTask(
+    stageId: Int,
```
Comment (JoshRosen) [inline streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala]: Style nit: these arguments should be indented another two spaces.

File: streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala [hunk 21]
```diff
@@ -295,6 +334,27 @@ class ReceiverSuite extends TestSuiteBase with Timeouts with Serializable {
...
+    executor: FakeReceiverSupervisor,
+    prefLocs: Seq[TaskLocation] = Nil) extends Task[Unit](stageId, 0) {
```
Comment (JoshRosen) [inline streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala]: It looks like we never call this with a non-Nil value of `prefLocs`, so we can probably drop the constructor argument and just hardcode `Nil` into the `preferredLocations` method.

File: streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala [hunk 22]
```diff
@@ -295,6 +334,27 @@ class ReceiverSuite extends TestSuiteBase with Timeouts with Serializable {
   }
 
   /**
+   * An implementation of Task running a NetworkReceiver used for testing whether
+   * we can start/stop receiver properly.
+   */
+  class FakeReceiverTask(
+    stageId: Int,
+    executor: FakeReceiverSupervisor,
```
Comment (JoshRosen) [inline streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala]: How about calling this `supervisor` instead of `executor`?

**Comment by JoshRosen** [comment]
On closer inspection, my [previous comment](https://github.com/apache/spark/pull/4135#discussion_r26727249) about testing still stands

> I notice that this test harness the logic from the real receiver tracker. Is this test suite actually covering the changes made to ReceiverTracker? In other words, if I were to go in and delete the callback registration code that you added to the real tracker, would that cause any tests to fail? If not, then I think we need to do additional work to test this, since the test isn't really testing anything if it can't expose bugs in the real code.

> It seems like we'd like to test against a real ReceiverTracker using mock implementations of ReceiverSupervisor and Receiver. I'm not super-familiar with this corner of the streaming code, though, so maybe @tdas can chime in with suggestions on how to test this?

Since the changes in this patch modify ReceiverTracker, it makes sense to test the actual ReceiverTracker while mocking other components.  The updated test still doesn't exercise the real receiver tracker and thus isn't actually covering the streaming changes in this patch.

I understand that properly testing ReceiverTracker might not be the easiest test to write, so perhaps this would require refactoring of ReceiverTracker to make the individual components more testable.  I'll let @tdas chime in to decide on the importance of writing a real test for this.  Sorry to be pedantic, but I'm a bit of a stickler for the importance of writing tests that actually exercise the code being modified, since I think it's very rare that it's truly impossible to test changes like this.

**Comment by uncleGen** [comment]
@JoshRosen Thanks for your careful and responsible review. I will hold this PR and wait for @tdas 's suggestions

**Comment by andrewor14** [comment]
@tdas @uncleGen any updates? If not, I would recommend that we close this PR for now and reopen it later if there is more activity.
